### ISSUE TITLE
library not found: stdc++

### ISSUE THREAD

**Issue by blasrodri** [rui314/mold#1084]
I'm trying to compile a repository: https://github.com/ComposableFi/composable

```
git clone https://github.com/ComposableFi/composable.git
cd code/
cargo b -p composable-runtime --release
```

Note that my config is
```
f1:~/composable/composable/code$ cat ~/.cargo/config.toml 
[target.x86_64-unknown-linux-gnu]
linker = "clang"
rustflags = ["-C", "link-arg=-fuse-ld=/usr/local/bin/mold"]

```

My error

```

  = note: mold: fatal: library not found: stdc++
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by rui314** [comment]
Can you build the program without mold?

**Comment by rui314** [comment]
Please reopen if it failed only with mold.

**Comment by blasrodri** [comment]
Sorry for the delay. I can confirm that it happens only while running this with mold.

**Comment by blasrodri** [comment]
@rui314 can you re-open it?

**Comment by rui314** [comment]
What is your distribution?

**Comment by blasrodri** [comment]
```
22.04.1-Ubuntu x86_64 x86_64 GNU/Linux
```

**Comment by ariesdevil** [comment]
Same issue for me when build [databend](https://github.com/datafuselabs/databend.git)
```
Linux ubuntu-linux-22-04-desktop 5.15.0-83-generic aarch64 aarch64 GNU/Linux
```

**Comment by dylanblokhuis** [comment]
Simply running ``sudo apt-get install libstdc++-12-dev`` fixed it for me

**Comment by sanmai-NL** [comment]
This fault should have nothing to do with `mold` ...

**Comment by Doods0** [comment]
As-salamu alaykum.

I encountered this issue while compiling a Rust program. This program compiles fine when I comment out these lines:
```
[target.x86_64-unknown-linux-gnu]
linker = "/usr/bin/clang"
rustflags = ["-C", "link-arg=-fuse-ld=mold"]
```

It is fixed by running this command:
```
sudo apt install libstdc++-14-dev
```
Notice it's version 14 and not 12 like a previous comment, probably because of the OS version.

I am running on `Pop!_OS 24.04`. It is essentially the same as `Ubuntu 24.04`.

Mold seems to have a weird link to `libstdc++`.

**Comment by knu** [comment]
This behavior stems from clang's attempt to detect and emulate GCC.  Clang searches for available GCC versions and automatically selects the most recent one to ensure compatibility with the system's standard library and toolchain.

For instance, Clang 18 on Ubuntu 24.04 scans for GCC versions from 9 up to 14, then selects the latest version found, which in this case is GCC 14.

```shellsession
% clang++ -v
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/12
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/13
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/14
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/14
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda, version
```

So, when Clang detects GCC 14 (even if it is not fully installed), it will not use libstdc++-13-dev (although it is the default version on Ubuntu 24.04); libstdc++-14-dev becomes required.

**Comment by janwilmans** [comment]
And with 'fully installed' you mean the -dev package is not installed? I have this issue with a version on Debian 12, with GCC-14.2 installed that I compiled myself, clang-19 installed and mold v2.40.3.

```
lib32stdc++6/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
**libstdc++-12-dev/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]**
libstdc++6/stable,bookworm,now 12.2.0-14+deb12u1 amd64 [installed,automatic]

gcc/stable,now 4:12.2.0-3 amd64 [installed,automatic]
gcc-12/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
gcc-12-base/stable,bookworm,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
**gcc-14.2.0/bookworm,now 14.2.0-3 amd64 [installed]**
lib32gcc-s1/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
libcc1-0/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
libgcc-12-dev/stable,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
libgcc-s1/stable,bookworm,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
libgomp1/stable,now 12.2.0-14+deb12u1 amd64 [installed]
libquadmath0/stable,bookworm,now 12.2.0-14+deb12u1 amd64 [installed,automatic]
```
without mold, it falls back to 'gold' and everything is fine, but when 'mold' is installed 

```
FAILED: src/infrastructure/image/image_test
```
: && /usr/bin/clang++-19 -O3 -DNDEBUG -fuse-ld=mold -no-pie -fsanitize=undefined /opt/fmt/lib/libfmt.a  src/infrastructure/image/libimage_library.a  src/infrastructure/toolkit/libtoolkit_library.a  /opt/gtest/lib/libgtest_main.a  /opt/gtest/lib/libgtest.a  /opt/fmt/lib/libfmt.a  /usr/lib/x86_64-linux-gnu/libpng.so  /usr/lib/x86_64-linux-gnu/libz.so && :
```
mold: fatal: library not found: stdc++
clang++-19: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by janwilmans** [comment]
This is the last part of the invocation:
```
 -lstdc++ -lm --no-as-needed -lpthread -lrt -lm -ldl -lresolv -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
mold: fatal: library not found: stdc++
```

It seems when 'mold' is passed '-lstdc++' it is unhappy about that.
In the invocation below I explicitly specified the path `/opt/gcc-14.2/lib64/libstdc++.a` and it starts complaining about the next -l argument


```
jan.wilmans@HAMILTON250{~/project/vcenter/build/release}: "/usr/bin/ld.mold" --hash-style=gnu --build-id --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o src/infrastructure/image/image_test /lib/x86_64-linux-gnu/crt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtbegin.o -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0 -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib --whole-archive /usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone-x86_64.a --no-whole-archive --dynamic-list=/usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone-x86_64.a.syms --whole-archive /usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone_cxx-x86_64.a --no-whole-archive --dynamic-list=/usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone_cxx-x86_64.a.syms src/infrastructure/image/CMakeFiles/image_test.dir/tests/copy_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/fill_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/makeimage_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/metadata_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/timestamp_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/png_pgm_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_test.cc.o /opt/fmt/lib/libfmt.a src/infrastructure/image/libimage_library.a src/infrastructure/toolkit/libtoolkit_library.a /opt/gtest/lib/libgtest_main.a /opt/gtest/lib/libgtest.a /opt/fmt/lib/libfmt.a /usr/lib/x86_64-linux-gnu/libpng.so /usr/lib/x86_64-linux-gnu/libz.so /opt/gcc-14.2/lib64/libstdc++.a -lm --no-as-needed -lpthread -lrt -lm -ldl -lresolv -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
```

the last part of the invocation above:
```
....
/opt/gtest/lib/libgtest.a /opt/fmt/lib/libfmt.a /usr/lib/x86_64-linux-gnu/libpng.so /usr/lib/x86_64-linux-gnu/libz.so /opt/gcc-14.2/lib64/libstdc++.a -lm --no-as-needed -lpthread -lrt -lm -ldl -lresolv -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
```

```
mold: fatal: library not found: gcc_s
```

**Comment by janwilmans** [comment]
I uninstalled 'mold' and re-ran the clang-19 invocation

```
"/usr/bin/ld" --hash-style=gnu --build-id --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o src/infrastructure/image/image_test /lib/x86_64-linux-gnu/crt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtbegin.o -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0 -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib --whole-archive /usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone-x86_64.a --no-whole-archive --dynamic-list=/usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone-x86_64.a.syms --whole-archive /usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone_cxx-x86_64.a --no-whole-archive --dynamic-list=/usr/lib/llvm-19/lib/clang/19/lib/linux/libclang_rt.ubsan_standalone_cxx-x86_64.a.syms src/infrastructure/image/CMakeFiles/image_test.dir/tests/copy_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/fill_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/makeimage_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/metadata_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/timestamp_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/png_pgm_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_test.cc.o /opt/fmt/lib/libfmt.a src/infrastructure/image/libimage_library.a src/infrastructure/toolkit/libtoolkit_library.a /opt/gtest/lib/libgtest_main.a /opt/gtest/lib/libgtest.a /opt/fmt/lib/libfmt.a /usr/lib/x86_64-linux-gnu/libpng.so /usr/lib/x86_64-linux-gnu/libz.so -lstdc++ -lm --no-as-needed -lpthread -lrt -lm -ldl -lresolv -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
```

I seems 'gold' is fine with '-lstdc++' but 'mold' is not, the invocation is not different.

**Comment by janwilmans** [comment]
In the scenario GCC14.2 + mold, I don't see this problem, its only clang-19 + mold.

```
gcc version 14.2.0 (GCC)
COMPILER_PATH=/opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/14.2.0/:/opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/14.2.0/:/opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/:/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/:/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/
LIBRARY_PATH=/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/:/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64/:/lib/x86_64-linux-gnu/:/lib/../lib64/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib64/:/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-O3' '-D' 'NDEBUG' '-fuse-ld=mold' '-fsanitize=undefined' '-o' 'src/infrastructure/image/image_test' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64' '-dumpdir' 'src/infrastructure/image/image_test.'
 /opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/14.2.0/collect2 -plugin /opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/14.2.0/liblto_plugin.so -plugin-opt=/opt/gcc-14.2/libexec/gcc/x86_64-linux-gnu/14.2.0/lto-wrapper -plugin-opt=-fresolution=/tmp/cc4pgzSw.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -fuse-ld=mold -o src/infrastructure/image/image_test /lib/x86_64-linux-gnu/crt1.o /lib/x86_64-linux-gnu/crti.o /opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/crtbegin.o -L/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0 -L/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/../../.. src/infrastructure/image/CMakeFiles/image_test.dir/tests/copy_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/fill_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/makeimage_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/metadata_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/timestamp_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/png_pgm_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_span_test.cc.o src/infrastructure/image/CMakeFiles/image_test.dir/tests/view_test.cc.o /opt/fmt/lib/libfmt.a src/infrastructure/image/libimage_library.a src/infrastructure/toolkit/libtoolkit_library.a /opt/gtest/lib/libgtest_main.a /opt/gtest/lib/libgtest.a /opt/fmt/lib/libfmt.a /usr/lib/x86_64-linux-gnu/libpng.so /usr/lib/x86_64-linux-gnu/libz.so -lstdc++ -lm -lubsan -lgcc_s -lgcc -lc -lgcc_s -lgcc /opt/gcc-14.2/lib/gcc/x86_64-linux-gnu/14.2.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
COLLECT_GCC_OPTIONS='-O3' '-D' 'NDEBUG' '-fuse-ld=mold' '-fsanitize=undefined' '-o' 'src/infrastructure/image/image_test' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64' '-dumpdir' 'src/infrastructure/image/image_test.'

```

This invocation in the same environment works without issue.

**Comment by janwilmans** [comment]
To reduce the noise, I re-ran the commands using 'main.c' containing only `int main() {return 0;}`

# using clang-19+mold:

```
jan.wilmans@HAMILTON250{~/project/vcenter/test}: clang-19 main.c -fuse-ld=mold
mold: fatal: library not found: gcc_s
clang-19: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
jan.wilmans@HAMILTON250{~/project/vcenter/test}: clang-19 main.c -fuse-ld=mold -v
Debian clang version 19.1.7 (++20250114103228+cd708029e0b2-1~exp1~20250114103334.78)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-19/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/14.2.0
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/14.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/usr/lib/llvm-19/bin/clang" -cc1 -triple x86_64-pc-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name main.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/home/jan.wilmans/project/vcenter/test -v -fcoverage-compilation-dir=/home/jan.wilmans/project/vcenter/test -resource-dir /usr/lib/llvm-19/lib/clang/19 -internal-isystem /usr/lib/llvm-19/lib/clang/19/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -ferror-limit 19 -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/main-e16c71.o -x c main.c
clang -cc1 version 19.1.7 based upon LLVM 19.1.7 default target x86_64-pc-linux-gnu
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../x86_64-linux-gnu/include"
ignoring nonexistent directory "/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/llvm-19/lib/clang/19/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
 "/usr/bin/ld.mold" --hash-style=gnu --build-id --eh-frame-hdr -m elf_x86_64 -pie -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /lib/x86_64-linux-gnu/Scrt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0 -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib /tmp/main-e16c71.o -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtendS.o /lib/x86_64-linux-gnu/crtn.o
mold: fatal: library not found: gcc_s
clang-19: error: linker command failed with exit code 1 (use -v to see invocation)
```



# using clang-19+gold:

```
jan.wilmans@HAMILTON250{~/project/vcenter/test}: clang-19 main.c -fuse-ld=gold
jan.wilmans@HAMILTON250{~/project/vcenter/test}:
```

```
jan.wilmans@HAMILTON250{~/project/vcenter/test}: clang-19 main.c -fuse-ld=gold -v
Debian clang version 19.1.7 (++20250114103228+cd708029e0b2-1~exp1~20250114103334.78)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-19/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/14.2.0
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/14.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/usr/lib/llvm-19/bin/clang" -cc1 -triple x86_64-pc-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name main.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/home/jan.wilmans/project/vcenter/test -v -fcoverage-compilation-dir=/home/jan.wilmans/project/vcenter/test -resource-dir /usr/lib/llvm-19/lib/clang/19 -internal-isystem /usr/lib/llvm-19/lib/clang/19/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -ferror-limit 19 -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/main-131332.o -x c main.c
clang -cc1 version 19.1.7 based upon LLVM 19.1.7 default target x86_64-pc-linux-gnu
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../x86_64-linux-gnu/include"
ignoring nonexistent directory "/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/llvm-19/lib/clang/19/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
 "/usr/bin/ld.gold" --hash-style=gnu --build-id --eh-frame-hdr -m elf_x86_64 -pie -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /lib/x86_64-linux-gnu/Scrt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0 -L/usr/lib/gcc/x86_64-linux-gnu/14.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib /tmp/main-131332.o -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-linux-gnu/14.2.0/crtendS.o /lib/x86_64-linux-gnu/crtn.o
```

**Comment by janwilmans** [comment]
@rui314 I would love to help resolve this, this happens only with mold, if I can help gather more information, please let me know, and please re-open this issue.

**Comment by rui314** [comment]
@janwilmans I believe your compiler does not pass the directory that contains the necessary libraries to the linker using the `-L` option. Unlike gold, mold does not include default library search paths, so every directory that may contain a library must be specified using `-L`.

I believe the gcc distributed as part of your distribution is configured that way. You may want to rebuild your gcc so that it behaves the same way.

Note that it is our explicit design choice not to hard-code default library paths into the linker. This is for build reproducibility and ease of cross-compilation. LLVM lld also doesn't have default search paths. We have no plans to change this.

**Comment by janwilmans** [comment]
That makes total sense. Any idea what kind of configuration gcc might need for this to happen? I will do some research on that myself tomorrow, but if you already have a clue that might save some time.

**Comment by rui314** [comment]
I actually don't know the name of the flag. Could you update the bug once you find out, for reference?

**Comment by PeterWrighten** [comment]
Same for me when i used clang++-15 to compile [sui](https://github.com/MystenLabs/sui) on rust...

UPDATE:

Solved by running `sudo apt install libstdc++-14-dev` on Ubuntu24.04LTS.

**Comment by janwilmans** [comment]
> [@janwilmans](https://github.com/janwilmans) I believe your compiler does not pass the directory that contains the necessary libraries to the linker using the `-L` option. Unlike gold, mold does not include default library search paths, so every directory that may contain a library must be specified using `-L`.
> 
> I believe the gcc distributed as part of your distribution is configured that way. You may want to rebuild your gcc so that it behaves the same way.
> 
> Note that it is our explicit design choice not to hard-code default library paths into the linker. This is for build reproducibility and ease of cross-compilation. LLVM lld also doesn't have default search paths. We have no plans to change this.

I misread you reply and went on a ghost-hunt, I tried to compare how gcc was build using `gcc -v` and `gcc-14` 

```
stock GCC 12 Bookworm (Debian 12)

Configured with: ../src/configure -v
--with-pkgversion='Debian 12.2.0-14+deb12u1'
--with-bugurl=file:///usr/share/doc/gcc-12/README.Bugs
--enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++,m2
--prefix=/usr
--with-gcc-major-version-only
--program-suffix=-12
--program-prefix=x86_64-linux-gnu-
--enable-shared
--enable-linker-build-id
--libexecdir=/usr/lib
--without-included-gettext
--enable-threads=posix
--libdir=/usr/lib
--enable-nls
--enable-clocale=gnu
--enable-libstdcxx-debug
--enable-libstdcxx-time=yes
--with-default-libstdcxx-abi=new
--enable-gnu-unique-object
--disable-vtable-verify
--enable-plugin
--enable-default-pie
--with-system-zlib
--enable-libphobos-checking=release
--with-target-system-zlib=auto
--enable-objc-gc=auto
--enable-multiarch
--disable-werror
--enable-cet
--with-arch-32=i686
--with-abi=m64
--with-multilib-list=m32,m64,mx32
--enable-multilib
--with-tune=generic
--enable-offload-targets=nvptx-none=/build/reproducible-path/gcc-12-12.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/reproducible-path/gcc-12-12.2.0/debian/tmp-gcn/usr
--enable-offload-defaulted
--without-cuda-driver
--enable-checking=release
--build=x86_64-linux-gnu
--host=x86_64-linux-gnu
--target=x86_64-linux-gnu

custom build GCC 14 Bookworm  (Debian 12)

Configured with: ../configure 
--build=x86_64-linux-gnu 
--host=x86_64-linux-gnu 
--target=x86_64-linux-gnu 
--enable-checking=release 
--enable-languages=c,c++,fortran 
--enable-libstdcxx-backtrace=yes
 --disable-multilib 
--program-suffix=-14.2 
--prefix /opt/gcc-14.2
```

However, it wasn't GCC at all that caused my problems. It was clang-19, as downloaded from: `https://apt.llvm.org/bookworm/pool/main/l/llvm-toolchain-19/` that gave me the issues, not anything I compiled myself.

So, I'm just reporting here, that from the above gcc configuration options only `--libdir=/usr/lib` seems to related to this issue, but for the standard library gcc generally behaves correctly by default, but clang-19 apparently does not.

**Comment by janwilmans** [comment]
As a final comment, summary:
 
- my system is a debian 12 installation
- it's default GCC is `/usr/bin/gcc-12`
- I build my own GCC, which is gcc-14.2.1, installed as `/usr/bin/gcc` 
- I installed a pre-build version of clang-19 from `https://apt.llvm.org/bookworm/pool/main/l/llvm-toolchain-19/`
- I build mold myself (mold 2.40.3), available as `/usr/bin/mold`

When building with gcc-14, mold gets past -L <path> with the correct path to find the gcc-14 libraries, I'm not sure why that works exactly, because I did not build gcc-14 it with any special options, (and not even with `--libdir=/usr/lib`), but I guess it makes sense that gcc would pass `-L <path>` to the linker pointing to its own libraries.

When building with clang-19, mold does not receive the -L <path>to the gcc-14 libraries (and that makes sense to me, because why should it?)

A fix for my specific case was to add: `-L/opt/gcc-14.2/lib64` to the linker step, however a better more general solution is to add `--gcc-toolchain=/opt/gcc-14.2 -stdlib=libstdc++` to the clang commandline. This lets clang know what to pass to mold as library directories.

In short: I was not telling clang what toolchain to use, so it was not able to pass the correct library directories to mold.
